### ISSUE TITLE
fieldCoordinates pointers not updated if mesh->coords_map is changed

### ISSUE THREAD

**Issue by johnomotani** [boutproject/BOUT-dev#1544]
`coords_map` in a `Mesh` may be changed, in particular if `Coordinates::geometry()` is called to update the `Coordinates` objects, which calls `Mesh::recalculateStaggeredCoordinates()`. This changes the pointers in `coords_map`, so any `Coordinates*` pointers that have previously been stored as `fieldCoordinates` in `Field`s become out of date. I think there are two possible solutions
1. in `recalculateStaggeredCoordinates()` create new `Coordinates` objects and then copy them into the existing objects in `coords_map`. Requires adding a `Coordinates::operator=(Coordinates)` method. This way the pointers in `coords_map` don't change, so existing `fieldCoordinates` variables are OK. For a test implementation see https://github.com/boutproject/BOUT-dev/tree/fix_coordinates_recalculation
2. don't store `fieldCoordinates` pointers in `Field`s. Instead look up using `fieldmesh->getCoordinates(getLocation())` each time `Field::getCoordinates()` is called. For test implementation see https://github.com/boutproject/BOUT-dev/tree/fix_coordinates_recalculation2

In both of the branches linked above, I've added checks in field operations, e.g. `ASSERT1(lhs.getCoordinates() == rhs.getCoordinates());` which make the bug throw an exception, for example in `tests/integrated/test-drift-instability-staggered` until the final, bugfix commit is applied.

I favour option 2 as a fix, because it also means we don't need to worry about updating `fieldCoordinates`, for example if the location of a `Field` is changed. There is a performance penalty from having the lookup from a `std::map` instead of storing the pointer, but I guess that would be small?

Whichever option we choose as the fix, I guess it should go into `master`. I'll make a PR with just the fix and not the extra checks when we decide.

**Comment by d7919** [comment]
We'd hope the performance penalty is small, but we do have a lot of these calls throughout the code. There's also a possible "race condition" (not in the usual sense but I can't think of the right phrase) in the second case (admittedly a bit contrived at the moment) as the "pass through" call to `getCoordinates` could in principle return a different pointer in two closely placed calls. For example something like

~~~~
ASSERT1(g.getCoordinates() == f.getCoordinates());
auto coords = f.getCoordinates();
doSomeOperationOn(f); //Oops this has a side effect of recalculating coordinates
ASSERT1(coords = g.getCoordinates()); // This test now fails
~~~~
This may also apply to the first option.

Once we've created a `Coordinates` for a particular location I think we should we just update that instance rather than throwing it away and creating a new one. I think this is probably mostly what (1) is proposing (that might be achieved with just a `swap` call if we're lucky?).

**Comment by johnomotani** [comment]
@d7919 that is a good point. There are `Coordinates* coord` pointers around in user code (like mine for example) that it would be better not to break.

> Once we've created a Coordinates for a particular location I think we should we just update that instance rather than throwing it away and creating a new one

Yes, that's the aim of option 1, and we could do it with a `swap` instead of `operator=` (implemented in `fix_coordinates_recalculation` now to check).

**Comment by johnomotani** [comment]
I don't think making the fields store `std::shared_ptr<Coordinates>` would help. If you make a copy of a `shared_ptr` and then update the original, the pointer stored in the second `shared_ptr` doesn't change (although its reference count might). So for example in this program `b` keeps the original value:
```
#include<memory>
#include<iostream>

int main() {
  auto a = std::make_shared<int>(1);
  
  auto b = a;

  auto c = std::make_shared<int>(2);

  std::cout<<"before swap"<<std::endl;
  std::cout<<"a "<<a<<" "<<*a<<std::endl;
  std::cout<<"b "<<b<<" "<<*b<<std::endl;
  std::cout<<"c "<<c<<" "<<*c<<std::endl;

  a.swap(c);
 
  std::cout<<std::endl<<"after swap"<<std::endl;
  std::cout<<"a "<<a<<" "<<*a<<std::endl;
  std::cout<<"b "<<b<<" "<<*b<<std::endl;
  std::cout<<"c "<<c<<" "<<*c<<std::endl;

  return 0;
}
```
the output is:
```
before swap
a 0x2574c30 1
b 0x2574c30 1
c 0x2574c50 2

after swap
a 0x2574c50 2
b 0x2574c30 1
c 0x2574c30 1
```
So I think we have to swap every member of Coordinates. An alternative would be to store pointers-to-pointers, but I don't think I like that.

**Comment by ZedThree** [comment]
A `weak_ptr` might be what we want: https://en.cppreference.com/w/cpp/memory/weak_ptr

We can then check if the reference still exists and if not, fetch the new one

**Comment by johnomotani** [comment]
There's another issue, basically what @d7919 said above: if in some user code there is
```
auto coords_ylow = mesh->getCoordinates(CELL_YLOW);
...
mesh->getCoordinates(CELL_CENTRE)->geometry();
//coords_ylow is now invalid
```
which makes me think now it is better to do as @d7919 suggested: once a `Coordinates` object is created, always keep the same object and just update it in-place.

**Comment by d7919** [comment]
(I think it is perhaps a philosophical question in some sense -- if a field is created with a certain coordinates should it keep a pointer to the original coordinates it was made with or should that silently update. Moving to storing a shared pointer allows us to follow both options. I suspect the most useful is the later approach.)

**Comment by d7919** [comment]
I guess the main issue with changing the type of pointer that we store is backwards compatibility. As we should use `getCoordinates` to get the pointer out of the field I think we could actually change the type of pointer we store and just convert this back to a raw ptr in the getCoordinates call. The change to the mesh interface may not be so easily handled?

**Comment by ZedThree** [comment]
It would be really nice if we did give out `shared_ptr`, I think that would solve a bunch of these problems. It would be a breaking change, but just replacing `Coordinates*` with `auto` would basically solve any breakages

**Comment by johnomotani** [comment]
If I understand right, giving out `shared_ptr` would mean any `coords` variable, e.g. from `auto coords = f.getCoordinates()` would point to the 'old' `Coordinates`, the state when `getCoordinates()` was called. I don't think that's what we want: if for some reason the `Coordinates` are changed/updated (which I assume would happen during initialization) then I think we always want to use the updated one.

**Comment by ZedThree** [comment]
Ah sorry, I didn't express that very clearly. We should:

1) Update the existing `Coordinates` in the map with the new values to keep everything in sync
2) Hand out `shared_ptr`s to not worry about ownership

**Comment by d7919** [comment]
I've pushed a commit to https://github.com/boutproject/BOUT-dev/tree/swap_coordinates_when_adding that I think should do what we want.

**Comment by ZedThree** [comment]
Fixed in #1549
